### DISCUSSION TITLE
Better Handling of Mime Types

### DISCUSSION THREAD

**Discussion by nighcoder** [clojupyter/clojupyter#153] [Ideas]
Currently, Clojupyter handles automatic rendering of data using a protocol, `PMimeConvertible` to attach the proper mime type to data in order to render properly on screen.

The protocol dispatches the proper mime tag based on the type of the object and this approach works great for data wrapped in a custom type.

Some types are already supported by default by Clojupyter:
* `java.awt.image.BufferedImage`
* the upcoming interactive widgets, a.k.a CommAtoms
* default types for **markdown**, **html** and **hiccup** data, used by render functions in display namespace.
* other types are printed as strings.

Other types can be extended by implementing the above protocol.

While this works for custom types, a lot of data we handle does not cast into a special type in Clojure world.
* **Vectors** can be **hiccup** data is just Clojure vectors and we have no way to differentiate them from the regular vectors (that aren't hiccup data).
* **Maps** can be a spec for **Vega** and **Vega lite**. These are both JSON maps that can be rendered by default in Jupyter Lab. We don't work with directly with JSON maps in Clojure, but we can serialize a subset of Clojure maps to JSON.
* **Strings** can be **plaintext**, **markdown**, **html** and even binary data encoded to a string.
* **Byte arrays** can also contain many different formats although they all have the same type.

I propose we start here a place to brainstorm how can we extend the current automatic rendering to richer data types, without expecting the end user to pass it to the right mime-renderer in display namespace.

Jupyter Lab can handle a few of these mime types without any plugins installed, and plugins can expand the list:
* `:text/plain` - What we ultimately use for most of the datatypes.
* `:text/html` - Produced by display/html from both html strings or hiccup data.
* `:text/markdown` - Produced by display/markdown.
* `:text/latex` - Produced by display/latex.
* `:image/bmp`, `:image/gif`, `:image/jpeg`, `:image/png`, `image/svg+xml`. `:image/png` is used by `clojupyter/*logo*` example in the readme.
* `:application/json`. Not used by Clojupyter.
* `:application/pdf`.
* `:application/vnd.vega.v5+json`
* `:application/vnd.vegalite.v3+json`
* `:application/vdom.v1+json`.

To cover some of the missing types, I've made the PR #124. It adds converters to generate the mimes types manually. Automatically, we have the BufferedImage from java.awt.image being rendered as `:image/png`, while the rest are printed as `:text/plain`.

How can we improve on that? Should we look into the data we're handling and choose a mime type?

Magic bytes can help us distinguish formats for binary data. Vega specs can be recognized by the `:$schema` key pointing to the Vega version. Strings are more difficult to deal with and distinguishing between plaintext, html and markdown can be expensive.

What are your thoughts on this? Is the MIME handler in need of an update, or it's fine as it is?

**Comment by behrica** [comment]
Please have a look at Clojure libraries kind/kindly which propose a renderer independent system for the same problem

**Reply by behrica** [reply]
In there we recognize that it is  not always possible to "detect" the format from the data.
The kindly libraries relies instead of "several approaches" in a concise library:
- use Clojure metadata to "tag" data
- use detection for certain types where this works
- potential tagging of data or "vars"

A UI independent "micro framework" which allows multiple potential "formats" per data objects (so called kinds)
https://github.com/scicloj/kindly
